File: Java/MergeSort.java
import java.util.*;

class MergeSort 
{

	private void merge(int[] a,int low,int high)
	{
		int mid=(low+high)/2;
		int i=low,j=mid+1;
		int[] c = new int[high-low+1];
		int k=0;
		while(i<=mid && j<=high)
		{
			if(a[i]<=a[j])
			{
				c[k]=a[i];
				i++;
				k++;
			}
			else
			{
				c[k]=a[j];
				j++;
				k++;
			}
		}
		while(i<=mid)
		{
			c[k]=a[i];
			k++;
			i++;
		}
		while(j<=high)
		{
			c[k]=a[j];
			k++;
			j++;
		}
		for(i=low;i<=high;i++)
		{
			a[i]=c[i-low];
		}
	}

	public void mergesort(int[] a,int low, int high)
	{
		if(low>=high)
			return;
		else
		{
			int mid=(low+high)/2;
			mergesort(a,low,mid);
			mergesort(a,mid+1,high);
			merge(a,low,high);
		}
	}

	public void print(int[] a,int n)
	{
		int i=0;
		for(i=0;i<n;i++)
			System.out.print(a[i]+" ");
	}

	public static void main(String[] args)
	{
		Scanner inputSize = new Scanner(System.in);
        System.out.println("Enter number of elements:");
        int size = inputSize.nextInt();

        Scanner inputElements = new Scanner(System.in);
        int[] a = new int[size];
        for (int i = 0; i < size; i++) {
            System.out.println("Enter the element " + (i + 1) + ":");
            a[i] = inputElements.nextInt();
        }
        inputSize.close();
        inputElements.close();

       	MergeSort mergeSort = new MergeSort();
       	mergeSort.mergesort(a,0,size-1);
       	mergeSort.print(a,size);
	}
}

File: Kotlin/MergeSort.kt
import java.util.*

class MergeSort {

    private fun merge(a: IntArray, low: Int, high: Int) {
        val mid = (low + high) / 2
        var i = low
        var j = mid + 1
        val c = IntArray(high - low + 1)
        var k = 0
        while (i <= mid && j <= high) {
            if (a[i] <= a[j]) {
                c[k] = a[i]
                i++
                k++
            } else {
                c[k] = a[j]
                j++
                k++
            }
        }
        while (i <= mid) {
            c[k] = a[i]
            k++
            i++
        }
        while (j <= high) {
            c[k] = a[j]
            k++
            j++
        }
        i = low
        while (i <= high) {
            a[i] = c[i - low]
            i++
        }
    }

    fun mergesort(a: IntArray, low: Int, high: Int) {
        if (low >= high)
            return
        else {
            val mid = (low + high) / 2
            mergesort(a, low, mid)
            mergesort(a, mid + 1, high)
            merge(a, low, high)
        }
    }

    fun print(a: IntArray, n: Int) {
        var i = 0
        i = 0
        while (i < n) {
            System.out.print(a[i].toString() + " ")
            i++
        }
    }

    companion object {

        fun main(args: Array<String>) {
            val inputSize = Scanner(System.`in`)
            System.out.println("Enter number of elements:")
            val size = inputSize.nextInt()

            val inputElements = Scanner(System.`in`)
            val a = IntArray(size)
            for (i in 0 until size) {
                System.out.println("Enter the element " + (i + 1) + ":")
                a[i] = inputElements.nextInt()
            }
            inputSize.close()
            inputElements.close()

            val mergeSort = MergeSort()
            mergeSort.mergesort(a, 0, size - 1)
            mergeSort.print(a, size)
        }
    }
}